### ISSUE TITLE
Wrong app icon in MATE and Xfce

### ISSUE THREAD

**Issue by clefebvre** [linuxmint/webapp-manager#44]
The webapp .desktop files define the following:

- StatupWMclass=XXX
- Exec=firefox --class XXX

Where XXX match. This should be enough to associate the windows to the .desktop file and it is therefore the responsibility of the the DE to do so and apply the right application icon. I'm not sure there's anything we can do here.

somebody showed me an SSB.sh generator script which circumvented that but it was really hacky.. it came with its own copy of Firefox Nightly and was basically overriding the actual Firefox icon file everytime a web app was launched.

We might need to patch MATE/Xfce to get this fixed properly.

**Comment by burhanuday** [comment]
Problem exists on Gnome 3 as well
Running Manjaro with Gnome DE

**Comment by prcr** [comment]
I have noticed an issue that perhaps is related to this one. Here's how to reproduce:

1. Create a Web App for Google Chrome and define a custom icon
2. Open Google Chrome normally
3. Open the newly created Web App, and it will have the default icon of Google Chrome.

I noticed just now that if the order of steps 2. and 3. is inverted, the Web App will have the custom icon defined and all subsequent Google Chrome windows will have that same custom icon from the Web App. :thinking: 

I'm using Linux Mint 19.3 with Cinnamon 4.4.8.

**Comment by JakubRimal** [comment]
I can confirm the behaviour which prcr described: all instances of Google Chrome have the same icon – an icon of the firstly launched instance is used for all instances launched later.

Note: In Firefox it works as expected.
Note 2: It works as expected when an isolated profile is used. But the use-case is to use the same profile as in the main browser window so I'm logged in, it has the same extensions etc.

**Comment by clefebvre** [comment]
In firefox it was fixed with https://github.com/linuxmint/webapp-manager/commit/dcf6620d6195c3ce6506c1b88a5f799a79b21a08

**Comment by clefebvre** [comment]
I can't reproduce the issue with Cinnamon here. I'll close this issue because it was primarily targeted at Firefox and MATE/Xfce.

I'll explain a little bit:

- Cinnamon uses appinfo to assign the icon in alt-tab and panel. So it doesn't really care what the window icon actually is, it uses the one defined in the .desktop file.
- MATE and Xfce don't do that, so they ended up with the wrong icons.
- We used a new feature in libxapp which uses a GTK3 module to force the window icon to the one specified in the .desktop file. Because Firefox is GTK3 we were able to fix Firefox icons in MATE/Xfce.
- This fix won't work for non-GTK browsers.
